karpenter-mpol/add-karpenter-nodeselector/add-karpenter-nodeselector.yaml

```diff
@@ -0,0 +1,69 @@
+apiVersion: policies.kyverno.io/v1alpha1
+kind: MutatingPolicy
+metadata:
+  name: add-karpenter-nodeselector
+  annotations:
+    policies.kyverno.io/title: Add Karpenter nodeSelector
+    policies.kyverno.io/category: Karpenter, EKS Best Practices
+    policies.kyverno.io/severity: medium
+    policies.kyverno.io/subject: Pod
+    policies.kyverno.io/minversion: 1.6.0
+    policies.kyverno.io/description: >- 
+      Selecting the correct Node(s) provisioned by Karpenter is a way to specify
+      the appropriate resource landing zone for a workload. This policy injects a
+      nodeSelector map into the Pod based on the Namespace type where it is deployed.
+spec:
+  evaluation:
+    admission:
+      enabled: true
+  matchConstraints:
+    resourceRules:
+    - apiGroups: [""]
+      apiVersions: ["v1"]
+      operations: ["CREATE", "UPDATE"]
+      resources: ["pods"]
+    namespaceSelector:
+      matchExpressions:
+      - key: type
+        operator: In
+        values: ["medium", "large"]
+  variables:
+    - name: namespaceLabels
+      expression: "resource.Get('v1', 'namespaces', '', object.metadata.namespace).metadata.labels"
+    - name: capacityType
+      expression: "variables.namespaceLabels.type == 'large' ? 'on-demand' : 'spot'"
+  mutations:
+  # Ensure nodeSelector exists
+  - patchType: JSONPatch
+    jsonPatch:
```

Comment:
Why use `jsonPatches`, not `applyConfiguration`? The limitation of json patch is that it doesn't support auto-gen for pod controllers.